### PR TITLE
Make relations visible to public user

### PR DESCRIPTION
**Pull request by gregorjerse**
if he has view permission on the respective collection.

### CODE REVIEW

File: resolwe/flow/tests/test_manager.py [hunk 1]
```diff
@@ -96,6 +99,17 @@ def test_workflow(self):
...
 
+    def test_communicate(self):
+        """Make sure communicate can serialize settings.
+
+        When object of PosixPath is used in settings the call to communicate
+        fails due to object not being JSON serializable by default.
+        """
+        with override_settings(FLOW_EXECUTOR={"DATA_DIR": Path("/some/path")}):
+            async_to_sync(manager.communicate)(run_sync=True)
+        # Make sure original settings are saved.
```
Comment (dblenkus) [inline resolwe/flow/tests/test_manager.py:110]: Are settings not saved by default in tests? If not, that can cause a lot of problems.

File: resolwe/flow/tests/test_relations.py [hunk 2]
```diff
@@ -469,9 +513,15 @@ def test_delete_superuser(self):
...
+    def test_delete_publicuser(self):
+        """Anonymous user must not be able to delete relations."""
+        resp = self._delete(self.relation_group.pk, user=AnonymousUser())
+        self.assertEqual(resp.status_code, status.HTTP_401_UNAUTHORIZED)
+        self.assertTrue(Relation.objects.filter(pk=self.relation_group.pk).exists())
+
     def test_delete_different_user(self):
         resp = self._delete(self.relation_group.pk, user=self.user)
-        self.assertEqual(resp.status_code, status.HTTP_404_NOT_FOUND)
+        self.assertEqual(resp.status_code, status.HTTP_401_UNAUTHORIZED)
```
Comment (dblenkus) [inline resolwe/flow/tests/test_relations.py:524]: Does the user have the view permission on the object? If not, 404 should be returned.

File: resolwe/flow/tests/test_manager.py (hunk 1 above)
Comment (gregorjerse) [reply resolwe/flow/tests/test_manager.py:110]: They are not saved by default when communicate is called from channels (like from listener, for example). So if original settings are not saved back (the last line), then all other tests fail.

File: resolwe/flow/tests/test_relations.py (hunk 2 above)
Comment (gregorjerse) [reply resolwe/flow/tests/test_relations.py:524]: He does not have the permission. I will change the returned code.

But is it not more appropriate to return unauthorized in this case? It is true that more information is revealed (the existence of object, for example) but I believe it is nice to return unauthorized when someone does not have the permission to view the object.

File: resolwe/flow/views/relation.py [hunk 3]
```diff
@@ -93,11 +93,18 @@ def update(self, request, *args, **kwargs):
         the collection referenced in the ``Relation``.
         """
         instance = self.get_object()
-        if (
-            not request.user.has_perm("edit_collection", instance.collection)
-            and not request.user.is_superuser
-        ):
-            return Response(status=status.HTTP_401_UNAUTHORIZED)
+
+        if not request.user.is_superuser:
```
Comment (dblenkus) [inline resolwe/flow/views/relation.py:97]: Put this check into a separate method and reuse it in the function below, e.g. something like `self.has_change_permission(instance, request.user)`.

**Review by dblenkus — COMMENTED** [review]
Looks good; I only have one comment. And don't forget to add a changelog.

File: resolwe/flow/tests/test_manager.py (hunk 1 above)
Comment (dblenkus) [reply resolwe/flow/tests/test_manager.py:110]: It would be great to force-update settings for each test. But that can be done in a separate pull request.

**Review by dblenkus — APPROVED** [review]

**Pull request merged**